<?xml version="1.0"?>
<launch>
  <!-- BEV-Planner Integration Launch File -->
  <!-- CARLA 환경에서 YOLOP + BEV-Planner 통합 시스템 실행 -->

  <!-- 파라미터 설정 -->
  <arg name="debug" default="false" doc="Enable debug mode with detailed logging"/>
  <arg name="use_gpu" default="true" doc="Use GPU for inference if available"/>
  <arg name="target_fps" default="30" doc="Target planning frequency"/>
  <arg name="namespace" default="bev_planner" doc="Node namespace"/>

  <!-- BEV-Planner 메인 노드 -->
  <node pkg="bev_planner_integration" 
        type="bev_planner_node.py" 
        name="bev_planner_node"
        ns="$(arg namespace)"
        output="screen"
        respawn="true">
    
    <!-- 기본 파라미터 -->
    <param name="debug" value="$(arg debug)"/>
    <param name="target_fps" value="$(arg target_fps)"/>
    <param name="use_gpu" value="$(arg use_gpu)"/>
    
    <!-- YOLOP 토픽 설정 -->
    <remap from="yolop_det_grid" to="/carla/yolop/det_grid"/>
    <remap from="yolop_da_grid" to="/carla/yolop/da_grid"/>
    <remap from="yolop_ll_grid" to="/carla/yolop/ll_grid"/>
    
    <!-- CARLA 토픽 설정 -->
    <remap from="ego_odometry" to="/carla/ego_vehicle/odometry"/>
    
    <!-- 출력 토픽 설정 -->
    <remap from="planned_trajectory" to="/$(arg namespace)/planned_trajectory"/>
    <remap from="debug_visualization" to="/$(arg namespace)/debug_vis"/>
    <remap from="statistics" to="/$(arg namespace)/statistics"/>
  </node>

  <!-- 궤적 시각화 노드 -->
  <node pkg="bev_planner_integration"
        type="trajectory_visualizer_node.py"
        name="trajectory_visualizer"
        ns="$(arg namespace)"
        output="screen">
    
    <!-- 입력 토픽 -->
    <remap from="planned_trajectory" to="/$(arg namespace)/planned_trajectory"/>
    <remap from="ego_odometry" to="/carla/ego_vehicle/odometry"/>
    
    <!-- 출력 토픽 -->
    <remap from="trajectory_markers" to="/$(arg namespace)/trajectory_markers"/>
    <remap from="safety_indicators" to="/$(arg namespace)/safety_indicators"/>
  </node>

  <!-- 성능 모니터링 노드 (디버그 모드에서만) - 현재 비활성화 -->
  <!--
  <node if="$(arg debug)" 
        pkg="bev_planner_integration"
        type="performance_monitor.py"
        name="performance_monitor"
        ns="$(arg namespace)" 
        output="screen">
    
    <remap from="statistics" to="/$(arg namespace)/statistics"/>
    <remap from="planned_trajectory" to="/$(arg namespace)/planned_trajectory"/>
  </node>
  -->

  <!-- RViz 시각화 (선택적) -->
  <arg name="rviz" default="false" doc="Launch RViz for visualization"/>
  <node if="$(arg rviz)" 
        pkg="rviz" 
        type="rviz" 
        name="rviz"
        output="screen"/>

  <!-- 로그 정보 -->
  <param name="/use_sim_time" value="false"/>
  
</launch> 